# ■ 자연어 처리 (NLP)
Natural Language Processing 이란

**자연어 처리(NLP, Natural Language Processing)**는 컴퓨터가 인간의 언어를 이해하고, 해석하고, 생성하는 기술을 말합니다. NLP는 언어학과 컴퓨터 과학의 융합 분야로, 언어를 처리하고 분석하는 다양한 작업을 포함합니다. 그 목적은 사람과 컴퓨터 간의 상호작용을 원활하게 만들어 주는 것입니다.

자연어 처리의 주요 작업은 다음과 같습니다:

1. **형태소 분석 (Morphological Analysis)**: 문장에서 단어를 나누고, 각 단어의 형태를 분석합니다. 예를 들어, 한국어에서 "사랑한다"라는 단어는 "사랑"과 "한다"로 나누어집니다.

1. **구문 분석 (Syntactic Parsing)**: 문장의 구조를 분석하고, 문법적 관계를 파악합니다. 예를 들어, "나는 사과를 먹는다"라는 문장에서 주어(나), 목적어(사과), 동사(먹는다) 관계를 식별합니다.

1. **의미 분석 (Semantic Analysis)**: 문장이 전달하려는 의미를 추론합니다. 예를 들어, "배가 고프다"는 단순히 '배'라는 단어가 아니라 '배'가 '허기짐'을 의미하는지를 이해합니다.

1. **감정 분석 (Sentiment Analysis)**: 텍스트에 포함된 감정을 분석하여 긍정적, 부정적, 중립적 감정을 분류합니다. 예를 들어, "이 영화는 정말 재미있다"는 긍정적인 감정을, "이 영화는 지루하다"는 부정적인 감정을 나타냅니다.

1. **기계 번역 (Machine Translation)**: 한 언어에서 다른 언어로 번역하는 작업입니다. 예를 들어, 영어에서 한국어로 번역하는 시스템을 개발하는 것입니다.

1. **질의 응답 (Question Answering)**: 사용자가 입력한 질문에 대해 적절한 답을 제공하는 시스템입니다. 예를 들어, "파리의 수도는 어디인가요?"라는 질문에 대해 "파리는 프랑스의 수도입니다."라는 답을 제공하는 것입니다.

1. **텍스트 생성 (Text Generation)**: 주어진 주제나 조건에 맞춰 텍스트를 자동으로 생성하는 작업입니다. 예를 들어, 사용자가 제공한 주제에 대해 자동으로 기사를 작성하는 것이 이에 해당합니다.

자연어 처리는 다양한 분야에서 활용됩니다. 예를 들어, 검색 엔진, 음성 인식 시스템, 챗봇, 자동 번역 시스템 등에서 사용되며, 최근에는 대규모 언어 모델(예: GPT) 등의 발전으로 자연어 처리 기술이 더욱 진화하고 있습니다.

# ■ 코퍼스 (corpus, 말뭉치)

자연어 분야의 연구를 위해 '특정한 목적'을 가지고 언어의 표본을 추출한 집합.

이를 사용하여 텍스트 데이터 분석및 저연어처리 (NLP) 에서 모델을 학습시키게 되는 데이터로 활용된다.

![](https://mblogthumb-phinf.pstatic.net/MjAyMjA1MDRfMTk4/MDAxNjUxNjI4NjQ4OTIx.ewwjMz43c_bhhXPLSv022-JqPZlMZmmjBh_P2_vSTfEg.y7dAYWl4kPg4Pm-30UIjKZoyiPriL6I0A9E8mjU83xQg.JPEG.mcstkorea/3._%EB%A7%90%EB%AD%89%EC%B9%98_%EC%A0%95%EC%9D%98.jpg?type=w800)


참조 사이트
국립국어원 말뭉치 검색
https://kcorpus.korean.go.kr/
![](https://img.sbs.co.kr/newsnet/test/sdf/20210927/1422164704/28394_1617005927.jpg)

# 설치패키지 & import

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import re

import tensorflow as tf
from tensorflow import keras

import torch
print('📍 torch version', torch.__version__)

# gensim: Generate Similar.
# 문서 간의 유사도(Similarity) 나 단어 의미 분석(Semantic Analysis) 에 초점을 둔
# 자연어 처리에 도움주는 패키지
!pip install gensim
import gensim

import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

# konlpy : 한국여형태소분석기 (+품사태깅)
!pip install konlpy
import konlpy


📍 torch version 2.8.0+cu126
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 70.8 MB/s eta 0:00:00


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.4/19.4 MB 17.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 495.9/495.9 kB 17.5 MB/s eta 0:00:00


# ■ 텍스트 전처리 (Text Preprocessing)

**텍스트 전처리**란?
주어진 문제를 해결하기 위해, 목적에 맞게 텍스트 데이터에 사전에 처리하는 작업입니다.

요리에 비유하자면, 아무리 조리방법과 장비가 좋다하더라도, 재료가 나쁘면, 혹은 재료가 제대로 손질되어 있지 않으면 좋은 요리가 나오기 힘듭니다.

텍스트 데이터 또한 적절한 전처리를 하지 않으면 앞으로 배울 자연어 처리 기법들이 제대로 동작하지 않습니다.

코퍼스 데이터는 사용하고자 하는 용도에 맞게 다양한 전처리 작업을 거치게 된다.

예를 들면.
- 토큰화(tokenization)
- 정제(cleaning)
- 정규화(normalization)


# ■ 02-01 토큰화(Tokenization)

- 주어진 코퍼스(corpus)에서 토큰(token)이라 불리는 단위로 나누는 작업을 토큰화(tokenization)라고 합니다

- 토큰의 단위는 상황에 따라 다르나, 보통 의미가 있는 단위로 토큰을 정의합니다.

## 🔹단어 토큰화(Word Tokenization)
토큰의 기준을 단어(word)로 하는 경우, 단어 토큰화(word tokenization)라고 합니다. 다만, 여기서 단어(word)는 단어 단위 외에도 단어구, 의미를 갖는 문자열로도 간주되기도 합니다.
예를 들어보겠습니다. 아래의 입력으로부터 구두점(punctuation)과 같은 문자는 제외시키는 간단한 단어 토큰화 작업을 해봅시다. 구두점이란 마침표(.), 컴마(,), 물음표(?), 세미콜론(;), 느낌표(!) 등과 같은 기호를 말합니다.


입력: **Time is an illusion. Lunchtime double so!**


이러한 입력으로부터 구두점을 제외시킨 토큰화 작업의 결과는 다음과 같습니다.


출력 : "Time", "is", "an", "illustion", "Lunchtime", "double", "so"

### NLTK 사용

In [2]:
sentence =  "Don't be fooled by the dark sounding name, Mr. Jone's Orphanage is as cheery as cheery goes for a pastry shop."

In [3]:
"""
Don't
Don t
Dont
Do n't
Jone's
Jone s
Jone
Jones
"""
None

In [4]:
nltk.download('punkt_tab')
from nltk.tokenize import word_tokenize
from nltk.tokenize import WordPunctTokenizer

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


In [5]:
print('단어 토큰화1:', word_tokenize(sentence))

단어 토큰화1: ['Do', "n't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'Mr.', 'Jone', "'s", 'Orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop', '.']


In [6]:
print('단어토큰화2:', WordPunctTokenizer().tokenize(sentence))

단어토큰화2: ['Don', "'", 't', 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'Mr', '.', 'Jone', "'", 's', 'Orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop', '.']


### keras: text_to_word_sequence()

In [7]:
# 케라스 또한 토큰화 도구로서 text_to_word_sequence를 지원합니다.
#   텍스트를 소문자화(lowercase)
#   지정된 구두점 제거
#   공백 기준으로 단어 단위 토큰화

In [8]:
from tensorflow.keras.preprocessing.text import text_to_word_sequence

In [9]:
print('단어토큰화3:', text_to_word_sequence(sentence))

단어토큰화3: ["don't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', 'mr', "jone's", 'orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop']


### re 사용

In [10]:
import re

def tokenize_sequence(text, filters=r"[!\"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n]", lower=True, split=" "):
  if lower:
    text = text.lower()

  text = re.sub(filters, "", text)
  return [word for word in text.split(split) if word]



In [11]:
print('단어토큰화4:', tokenize_sequence(sentence))

단어토큰화4: ["don't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name,', 'mr.', "jone's", 'orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop.']


### HuggingFace Tokenizer / Transformer  활용

In [12]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')  # 사전학습모델 다운로드

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [13]:
print('단어토큰화5', tokenizer.tokenize(sentence))

단어토큰화5 ['don', "'", 't', 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'mr', '.', 'jon', '##e', "'", 's', 'orphanage', 'is', 'as', 'cheer', '##y', 'as', 'cheer', '##y', 'goes', 'for', 'a', 'pastry', 'shop', '.']


## 토큰화에서 고려해야 할 사항

### 구두점과 특수문자
구두점과 특수문자를 무조건 제외하는게 답은 아니다

In [14]:
# 예]
#   단어 자체에 구두점 : m.p.h 나 Ph.D 나 AT&T
#   $, / 같은 특수문자 : $45.55 나 01/02/06
#   숫자 사이의 컴마: 123,456,789

### 띄어쓰기
줄임말과 단어 내에 띄어쓰기가 있는 경우.

In [15]:
# 예]
#  what're는 what are의 줄임말, we're는 we are의 줄임말

#  "New York" 이나 "rock 'n' roll" 는 띄어쓰기가 있지만 한 단어로 토큰화 할  필요도 있다.


## 표준 토큰화 예제

In [16]:
# Penn Treebank Tokenization의 규칙

# 규칙 1. 하이픈으로 구성된 단어는 하나로 유지한다.
# 규칙 2. doesn't와 같이 아포스트로피로 '접어'가 함께하는 단어는 분리해준다.


In [17]:
text = "Starting a home-based restaurant may be an ideal. it doesn't have a food chain or restaurant of their own."

In [18]:
from nltk.tokenize import TreebankWordTokenizer

In [19]:
tokenizer = TreebankWordTokenizer()

print('트리뱅크 워드코트나이저:', tokenizer.tokenize(text))

트리뱅크 워드코트나이저: ['Starting', 'a', 'home-based', 'restaurant', 'may', 'be', 'an', 'ideal.', 'it', 'does', "n't", 'have', 'a', 'food', 'chain', 'or', 'restaurant', 'of', 'their', 'own', '.']


## 4. 🔹문장 토큰화(Sentence Tokenization)
- 토큰의 단위가 문장(sentence) 일 경우.

In [20]:
#  !, ? 는 문장구분을 위한 명확한 구분자 (boundary) 역할을 하지만
# 마침표(.) 는 그렇지 않다

In [21]:
# EX1)
# IP 192.168.56.31 서버에 들어가서 로그 파일 저장해서 aaa@gmail.com로 결과 좀 보내줘. 그 후 점심 먹으러 가자.

# EX2)
# Since I'm actively looking for Ph.D. students, I get the same question a dozen times every year.


### NLTK sent_tokenize()

In [22]:
from nltk.tokenize import sent_tokenize

In [23]:
text = "His barber kept his word. But keeping such a huge secret to himself was driving him crazy. Finally, the barber went up a mountain and almost to the edge of a cliff. He dug a hole in the midst of some reeds. He looked about, to make sure no one was near."
text

'His barber kept his word. But keeping such a huge secret to himself was driving him crazy. Finally, the barber went up a mountain and almost to the edge of a cliff. He dug a hole in the midst of some reeds. He looked about, to make sure no one was near.'

In [24]:
print(sent_tokenize(text))

['His barber kept his word.', 'But keeping such a huge secret to himself was driving him crazy.', 'Finally, the barber went up a mountain and almost to the edge of a cliff.', 'He dug a hole in the midst of some reeds.', 'He looked about, to make sure no one was near.']


In [25]:
text = "I am actively looking for Ph.D. students. and you are a Ph.D student."
print(sent_tokenize(text))

['I am actively looking for Ph.D. students.', 'and you are a Ph.D student.']


### kss

한국어에 대한 문장 토큰화 도구 또한 존재합니다. 한국어의 경우에는 박상길님이 개발한 KSS(Korean Sentence Splitter)를 추천합니다. 다음과 같이 KSS를 설치합니다.

In [26]:
!pip install kss

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 163.0/163.0 kB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.4/26.4 MB 54.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 180.3/180.3 kB 12.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 19.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.3/131.3 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 939.7/939.7 kB 19.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.2/72.2 MB 9.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 50.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.3/103.3 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.8/235.8 kB 11.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.8/468.8 kB 27.4 MB/s eta 0:00:00

In [27]:
import kss

In [28]:
text = '딥 러닝 자연어 처리가 재미있기는 합니다. 그런데 문제는 영어보다 한국어로 할 때 너무 어렵습니다. 이제 해보면 알걸요?'


In [29]:
print(kss.split_sentences(text))

For your information, Kss also supports mecab backend.
We recommend you to install mecab or konlpy.tag.Mecab for faster execution of Kss.
Please refer to following web sites for details:
- mecab: https://github.com/hyunwoongko/python-mecab-kor
- konlpy.tag.Mecab: https://konlpy.org/en/latest/api/konlpy.tag/#mecab-class

/usr/local/lib/python3.12/dist-packages/pecab/_tokenizer.py:265: RuntimeWarning: overflow encountered in scalar add
  from_pos_data.costs[idx]


['딥 러닝 자연어 처리가 재미있기는 합니다.', '그런데 문제는 영어보다 한국어로 할 때 너무 어렵습니다.', '이제 해보면 알걸요?']


## 한국어 토큰화는 더 어렵다

In [30]:
# 교착어: 조사, 어미 등을 붙여서 말을 만드는 언어.


### 이유1) 한국어는 교착어이기 때문에

교착어의 의 특성
1. 조사
1. 띄어쓰기 단위
1. 형태소 (morpheme)

In [31]:
# 영어 '그' he/him 정도 뿐일텐데
#  한국어는  '그가', '그에게', '그를',  '그와',  '그는' ... 다양한 조사가 붙는다.

# ★ 대부분의 한국어 NLP 에선 조사는 분리해줄 필요가 있다.

# 한국어 토큰화 --> '형태소' 란 개념 반드시 이해
# 형태소(morpheme) : 뜻을 가진 가장 작은 말의 단위

In [32]:
# 문장 : 에디가 책을 읽었다

# ▶이 문장을 띄어쓰기 단위 토큰화를 수행한다면
#   ['에디가', '책을', '읽었다']

# ▶ 형태소 단위로 분해하면
#   자립 형태소 : 에디, 책
#   의존 형태소 : -가, -을, 읽-, -었, -다

### 이유2) 한국어의 띄어쓰기는 영어보다 잘 지켜지지 않기 때문에

In [33]:
# EX1) 제가이렇게띄어쓰기를전혀하지않고글을썼다고하더라도글을이해할수있습니다.

# EX2) Tobeornottobethatisthequestion


## 품사 태깅 (Part-Of-Speech tagging)

In [34]:
# 'fly'  명사 - 파리
# 'fly'  동사 - 날다

# '못'   명사 - 못 박는다.
# '못'   부사 - 못한다.

# 단어 토큰화 과정에서 각 단어가 어떤 품사로 쓰였는지를 구분해놓기도 하는데,
# 이 작업을 품사 태깅(part-of-speech tagging)이라고 합니다.

### NLTK 영어 품사태깅

In [35]:
from nltk.tag import pos_tag

In [36]:
nltk.download('averaged_perceptron_tagger_eng')

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger_eng.zip.


True

In [37]:
text = "I am actively looking for Ph.D. students. and you are a Ph.D. student."
tokenized_sentence = word_tokenize(text)
print('단어 토큰화 :',tokenized_sentence)

단어 토큰화 : ['I', 'am', 'actively', 'looking', 'for', 'Ph.D.', 'students', '.', 'and', 'you', 'are', 'a', 'Ph.D.', 'student', '.']


In [38]:
print('품사태깅: ', pos_tag(tokenized_sentence))

품사태깅:  [('I', 'PRP'), ('am', 'VBP'), ('actively', 'RB'), ('looking', 'VBG'), ('for', 'IN'), ('Ph.D.', 'NNP'), ('students', 'NNS'), ('.', '.'), ('and', 'CC'), ('you', 'PRP'), ('are', 'VBP'), ('a', 'DT'), ('Ph.D.', 'NNP'), ('student', 'NN'), ('.', '.')]


In [39]:
# Penn Treebank POG Tags에서 PRP는 인칭 대명사, VBP는 동사, RB는 부사, VBG는 현재부사,
# IN은 전치사, NNP는 고유 명사, NNS는 복수형 명사, CC는 접속사, DT는 관사를 의미합니다.


### KoNLPy 한국어 품사태깅

In [40]:
# KoNLPy  코엔엘파이 http://konlpy.org/ko/latest/

# 코엔엘파이를 통해서 사용할 수 있는 형태소 분석기로 Okt(Open Korea Text), 메캅(Mecab),
# 코모란(Komoran), 한나눔(Hannanum), 꼬꼬마(Kkma)가 있습니다.


In [41]:
from konlpy.tag import Okt
from konlpy.tag import Kkma

In [42]:
okt = Okt()
kkma = Kkma()

In [43]:
text = "열심히 코딩한 당신, 연휴에는 여행을 가봐요"

In [44]:
print('OKT 형태소 분석:', okt.morphs(text))
print('OKT 품사태깅:', okt.pos(text))
print('OKT 명사추출:', okt.nouns(text))

OKT 형태소 분석: ['열심히', '코딩', '한', '당신', ',', '연휴', '에는', '여행', '을', '가봐요']
OKT 품사태깅: [('열심히', 'Adverb'), ('코딩', 'Noun'), ('한', 'Josa'), ('당신', 'Noun'), (',', 'Punctuation'), ('연휴', 'Noun'), ('에는', 'Josa'), ('여행', 'Noun'), ('을', 'Josa'), ('가봐요', 'Verb')]
OKT 명사추출: ['코딩', '당신', '연휴', '여행']


In [45]:
print('꼬꼬마 형태소 분석:', kkma.morphs(text))
print('꼬꼬마 품사태깅:', kkma.pos(text))
print('꼬꼬마 명사추출:', kkma.nouns(text))

꼬꼬마 형태소 분석: ['열심히', '코딩', '하', 'ㄴ', '당신', ',', '연휴', '에', '는', '여행', '을', '가보', '아요']
꼬꼬마 품사태깅: [('열심히', 'MAG'), ('코딩', 'NNG'), ('하', 'XSV'), ('ㄴ', 'ETD'), ('당신', 'NP'), (',', 'SP'), ('연휴', 'NNG'), ('에', 'JKM'), ('는', 'JX'), ('여행', 'NNG'), ('을', 'JKO'), ('가보', 'VV'), ('아요', 'EFN')]
꼬꼬마 명사추출: ['코딩', '당신', '연휴', '여행']


---
# ■ 02-02 정제(Cleaning), 정규화(Normalization)

- **정제(cleaning)** : 갖고 있는 코퍼스로부터 노이즈 데이터를 제거한다.
- **정규화(normalization)** : 표현 방법이 다른 단어들을 통합시켜서 같은 단어로 만들어준다.

In [46]:
# 정제 및 정규화는 '토큰화 작업 전 (혹은 후)' 에 수행하곤 한다.

## 규칙에 기반한 표기가 다른 단어들의 통합

In [47]:
# ex)  USA 와 US,  uh-huh 와 uhhuh <= 형태는 다르지만 여전히 같은 의미

# 표기가 다른 단어들을 통합하는 방법중
#   - 어간추출 (stemming)
#   - 표제어추출(lemmatization)

## 대, 소문자 통합
- 알파벳 언어의 또 다른 정규화 방법

In [48]:
# 대소문자를 통합해야 하는 경우
# ex) Ferrari 로 검색했을때와 ferrari 로 검색했을때 동일한 결과를 얻을수 있게 해야 한다

# 대소문자가 구분시켜야 하는 경우
# ex) US 와 us 는 다른 의미의 단어.
# ex) 회사 이름(General Motors)나, 사람 이름(Bush) 등은 대문자로 유지

## 불용어(Stop word) 제거
불필요한 단어 제거

ex)
1. 등장빈도가 적은 단어
1. 길이가 짧은 단어

In [49]:
text = "I was wondering if anyone out there could enlighten me on this car."

# 길이가 1~2 인 단어들을 삭제
shortword = re.compile(r'\W*\b\w{1,2}\b')
print(shortword.sub('', text))


 was wondering anyone out there could enlighten this car.


## 정규표현식

In [50]:
# 노이즈 문자들 제거용 ex) HTML 문서내의 tag들 <div> <a>....

# ■ 02-03 어간 추출(Stemming) , 표제어 추출(Lemmatization)

In [51]:
# 이 두 작업이 갖고 있는 의미는. 눈으로 봤을 때는 서로 다른 단어들이지만,
# 하나의 단어로 일반화시킬 수 있다면 하나의 단어로 일반화시켜서
# 문서 내의 단어 수(종류)를 줄이겠다는 거다


## 표제어 추출 (Lemmatization)

In [52]:
# 표제어(Lemma) : '기본 사전형 단어'
# 표제어 추출은 단어들이 다른 형태를 가지더라도, 그 뿌리 단어를 찾아가서 단어의 개수를 줄일 수 있는지 판단
# ex) am, are, is는 서로 다른 스펠링이지만 그 뿌리 단어는 be

# 형태소: '의미를 가진 가장 작은 단위'
# 형태소의 종류: 어간(stem)과 접사(affix)

# 1) 어간(stem)
# : 단어의 의미를 담고 있는 단어의 핵심 부분.

# 2) 접사(affix)
# : 단어에 추가적인 의미를 주는 부분.


In [53]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [54]:
from nltk.stem import WordNetLemmatizer

In [55]:
lemmatizer = WordNetLemmatizer()

In [56]:
words = ['policy', 'doing', 'organization', 'have', 'going', 'love', 'lives', 'fly', 'dies', 'watched', 'has', 'starting']

print('표제어 추출 전 :',words)
print('표제어 추출 후 :', [lemmatizer.lemmatize(word) for word in words])




표제어 추출 전 : ['policy', 'doing', 'organization', 'have', 'going', 'love', 'lives', 'fly', 'dies', 'watched', 'has', 'starting']
표제어 추출 후 : ['policy', 'doing', 'organization', 'have', 'going', 'love', 'life', 'fly', 'dy', 'watched', 'ha', 'starting']


In [57]:
# 위의 결과에서는 dy나 ha와 같이 의미를 알 수 없는 적절하지 못한 단어를 출력하고 있습니다.
# 이는 표제어 추출기(lemmatizer)가 본래 단어의 품사 정보를 알아야만 정확한 결과를 얻을 수 있기 때문입니다.


In [58]:
# WordNetLemmatizer는 입력으로 단어가 동사 품사라는 사실을 알려줄 수 있습니다.
# 즉, dies와 watched, has가 문장에서 동사로 쓰였다는 것을 알려준다면
# 표제어 추출기는 품사의 정보를 보존하면서 정확한 Lemma를 출력하게 됩니다.


In [59]:
lemmatizer.lemmatize('dies', 'v')

'die'

In [60]:
lemmatizer.lemmatize('watched', 'v')

'watch'

In [61]:
lemmatizer.lemmatize('has', 'v')

'have'

## 어간 추출 (Stemming)

In [62]:
# 어간(Stem)을 추출하는 작업을 어간 추출(stemming)이라고 합니다.
# 어간 추출은 형태학적 분석을 단순화한 버전이라고 볼 수도 있고,
# 정해진 규칙만 보고 단어의 어미를 자르는 어림짐작의 작업이라고 볼 수도 있다.


In [63]:
sentence = "This was not the map we found in Billy Bones's chest, but an accurate copy, complete in all things--names and heights and soundings--with the single exception of the red crosses and the written notes."

In [64]:
from nltk.stem import PorterStemmer

In [65]:
stemmer = PorterStemmer()

In [66]:
tokenized_sentence = word_tokenize(sentence)

print('어간 추출 전:', tokenized_sentence)
print('어간 추출 후:', [stemmer.stem(word) for word in tokenized_sentence])

어간 추출 전: ['This', 'was', 'not', 'the', 'map', 'we', 'found', 'in', 'Billy', 'Bones', "'s", 'chest', ',', 'but', 'an', 'accurate', 'copy', ',', 'complete', 'in', 'all', 'things', '--', 'names', 'and', 'heights', 'and', 'soundings', '--', 'with', 'the', 'single', 'exception', 'of', 'the', 'red', 'crosses', 'and', 'the', 'written', 'notes', '.']
어간 추출 후: ['thi', 'wa', 'not', 'the', 'map', 'we', 'found', 'in', 'billi', 'bone', "'s", 'chest', ',', 'but', 'an', 'accur', 'copi', ',', 'complet', 'in', 'all', 'thing', '--', 'name', 'and', 'height', 'and', 'sound', '--', 'with', 'the', 'singl', 'except', 'of', 'the', 'red', 'cross', 'and', 'the', 'written', 'note', '.']


In [67]:
# '규칙 기반'의 접근을 하고 있으므로 어간 추출 후의 결과에는 사전에 없는 단어들도 포함되어 있습니다.
# 가령, 포터 알고리즘의 어간 추출은 이러한 규칙들을 가집니다.

# ALIZE → AL
# ANCE → 제거
# ICAL → IC

# 위의 규칙에 따르면 좌측의 단어는 우측의 단어와 같은 결과를 얻게됩니다.

# formalize → formal
# allowance → allow
# electricical → electric


In [68]:
words = ['formalize', 'allowance', 'electricical']

print('어간 추출 전 :',words)
print('어간 추출 후 :',[stemmer.stem(word) for word in words])

어간 추출 전 : ['formalize', 'allowance', 'electricical']
어간 추출 후 : ['formal', 'allow', 'electric']


In [69]:
from nltk.stem import PorterStemmer
from nltk.stem import LancasterStemmer

porter_stemmer = PorterStemmer()
lancaster_stemmer = LancasterStemmer()

words = ['policy', 'doing', 'organization', 'have', 'going', 'love', 'lives', 'fly', 'dies', 'watched', 'has', 'starting']
print('어간 추출 전 :', words)
print('포터 스테머의 어간 추출 후:',[porter_stemmer.stem(w) for w in words])
print('랭커스터 스테머의 어간 추출 후:',[lancaster_stemmer.stem(w) for w in words])

어간 추출 전 : ['policy', 'doing', 'organization', 'have', 'going', 'love', 'lives', 'fly', 'dies', 'watched', 'has', 'starting']
포터 스테머의 어간 추출 후: ['polici', 'do', 'organ', 'have', 'go', 'love', 'live', 'fli', 'die', 'watch', 'ha', 'start']
랭커스터 스테머의 어간 추출 후: ['policy', 'doing', 'org', 'hav', 'going', 'lov', 'liv', 'fly', 'die', 'watch', 'has', 'start']


In [70]:
# 동일한 단어에 대해서 표제어 추출과 어간 추출을 각각 수행했을 때,
# 결과 차이 예

# Stemming
# am → am
# the going → the go
# having → hav

# Lemmatization
# am → be
# the going → the going
# having → have

## 한국어에서의 어간 추출

- 한국어는 5언 9품사 구조

|언|품사|
|------|---|
|체언|명사, 대명사, 수사|
|수식언|관형사, 부사|
|관계언|조사|
|독립언|감탄사|
|**용언**|**동사, 형용사**|

- 용언에 해당되는 '동사'와 '형용사'는 어간(stem)과 어미(ending)의 결합으로 구성

In [71]:
#  규칙 활용
#   잡다  => '잡' / 어간  + '다' / 어미

In [72]:
# 불규칙활용

# '듣다' => '듣/들-',
# '돕다' => '돕/도우-'
# '곱다' => '곱/고우-'

# ■ 02-04 불용어(Stopword)

In [73]:
# 갖고 있는 데이터에서 유의미한 단어 토큰만을 선별하기 위해서는
# 큰 의미가 없는 단어 토큰을 제거해야 한다

# 예를 들면, I, my, me, over, 조사, 접미사 같은 단어들은 문장에서는 자주 등장하지만
# 실제 의미 분석을 하는데는 거의 기여하는 바가 없는 경우가 있습니다.
# 이러한 단어들을 불용어(stopword)라고 하며,
# NLTK에서는 위와 같은 100여개 이상의 영어 단어들을 불용어로 패키지 내에서 미리 정의하고 있습니다.


In [74]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [75]:
from nltk.corpus import stopwords

## NLTK 에서 불용어 확인하기

In [76]:
stop_words_list = stopwords.words('english')
print('불용어 개수:', len(stop_words_list))
print('불용어 10개 출력:', stop_words_list[:10])

불용어 개수: 198
불용어 10개 출력: ['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an']


## NLTK 에서 불용어 제거하기

In [77]:
example = "Family is not an important thing. It's everything."
stop_words = set(stopwords.words('english'))

word_tokens = word_tokenize(example)

result = []

for word in word_tokens:
  if word not in stop_words:
    result.append(word)

print('불용어 제거 전:', word_tokens)
print('불용어 제거 후:', result)


불용어 제거 전: ['Family', 'is', 'not', 'an', 'important', 'thing', '.', 'It', "'s", 'everything', '.']
불용어 제거 후: ['Family', 'important', 'thing', '.', 'It', "'s", 'everything', '.']


## 한국어에서 불용어 제거

In [78]:
# 방법1: 간단하게 '조사' '접속사' 제거하는 방법
# 방법2: '사용자 지정 불용어 사전' 제공하여 제거하는 방법


In [79]:
example = "고기를 아무렇게나 구우려고 하면 안 돼. 고기라고 다 같은 게 아니거든. 예컨대 삼겹살을 구울 때는 중요한 게 있지."
stop_words = "를 아무렇게나 구 우려 고 안 돼 같은 게 구울 때 는"

stop_words = set(stop_words.split())
word_tokens = okt.morphs(example)

result = [
    word
    for word in word_tokens
    if word not in stop_words
]

print('불용어 제거 전:', word_tokens)
print('불용어 제거 후:', result)





불용어 제거 전: ['고기', '를', '아무렇게나', '구', '우려', '고', '하면', '안', '돼', '.', '고기', '라고', '다', '같은', '게', '아니거든', '.', '예컨대', '삼겹살', '을', '구울', '때', '는', '중요한', '게', '있지', '.']
불용어 제거 후: ['고기', '하면', '.', '고기', '라고', '다', '아니거든', '.', '예컨대', '삼겹살', '을', '중요한', '있지', '.']


In [80]:
# 아래의 링크는 보편적으로 선택할 수 있는 한국어 불용어 리스트를 보여줍니다.
# 하지만 여전히 절대적인 기준은 아닙니다.

# 링크 : https://www.ranks.nl/stopwords/korean

# 불용어가 많은 경우에는 코드 내에서 직접 정의하지 않고 txt 파일이나 csv 파일로 정리해놓고
# 이를 불러와서 사용하기도 합니다.

# ■ 02-06 정수 인코딩(Integer Encoding)
텍스트 -> 숫자

- 각 단어를 '고유한 정수' 에 맵핑(mapping) 시키는 전처리 작업

In [81]:
# 단어 인덱스 부여

# 일반적으로 단어의 등장 빈도수를 기준으로 정렬한뒤 부여함.

## 정수 인코딩 (Integer Encoding)

### 1) dict 를 사용하여 인코딩

In [82]:
import nltk
nltk.download('stopwords')
nltk.download('punkt_tab')
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [83]:
raw_text = "A barber is a person. a barber is good person. a barber is huge person. he Knew A Secret! The Secret He Kept is huge secret. Huge secret. His barber kept his word. a barber kept his word. His barber kept his secret. But keeping and keeping such a huge secret to himself was driving the barber crazy. the barber went up a huge mountain."

In [84]:
sentences = sent_tokenize(raw_text)
sentences

['A barber is a person.',
 'a barber is good person.',
 'a barber is huge person.',
 'he Knew A Secret!',
 'The Secret He Kept is huge secret.',
 'Huge secret.',
 'His barber kept his word.',
 'a barber kept his word.',
 'His barber kept his secret.',
 'But keeping and keeping such a huge secret to himself was driving the barber crazy.',
 'the barber went up a huge mountain.']

In [85]:
vocab = {}  # 각 단어별 빈도수
preprocessed_sentences = []
stop_words = set(stopwords.words('english'))

for sentence in sentences:
  # 각 문장마다 단어 토큰화
  tokenized_sentence = word_tokenize(sentence)

  result = []  # 전처리된 문장의 단어들

  for word in tokenized_sentence:
    word = word.lower()  # 모든 단어를 소문자화 하여 단어의 개수를 줄인다.
    if word not in stop_words:  # 단어 토큰중 불용어는 제거
      if len(word) > 2:    # 단어 길이가 2 이하인 경우 제거.
        result.append(word)
        # 빈도수 체크
        if word not in vocab:
          vocab[word] = 0
        vocab[word] += 1


  preprocessed_sentences.append(result)


preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [86]:
# 단어빈도수 (단어집합)
print(vocab)
vocab

{'barber': 8, 'person': 3, 'good': 1, 'huge': 5, 'knew': 1, 'secret': 6, 'kept': 4, 'word': 2, 'keeping': 2, 'driving': 1, 'crazy': 1, 'went': 1, 'mountain': 1}


{'barber': 8,
 'person': 3,
 'good': 1,
 'huge': 5,
 'knew': 1,
 'secret': 6,
 'kept': 4,
 'word': 2,
 'keeping': 2,
 'driving': 1,
 'crazy': 1,
 'went': 1,
 'mountain': 1}

In [87]:
vocab['barber']

8

In [88]:
# 빈도수가 높은 순서대로 정렬

vocab_sorted = sorted(vocab.items(), key=lambda x:x[1], reverse=True)
print(vocab_sorted)

vocab_sorted

[('barber', 8), ('secret', 6), ('huge', 5), ('kept', 4), ('person', 3), ('word', 2), ('keeping', 2), ('good', 1), ('knew', 1), ('driving', 1), ('crazy', 1), ('went', 1), ('mountain', 1)]


[('barber', 8),
 ('secret', 6),
 ('huge', 5),
 ('kept', 4),
 ('person', 3),
 ('word', 2),
 ('keeping', 2),
 ('good', 1),
 ('knew', 1),
 ('driving', 1),
 ('crazy', 1),
 ('went', 1),
 ('mountain', 1)]

In [89]:
# 높은 빈도수를 가진 단어에 낮은 정수를 부여.  정수 1부터 부여.

word_to_index = {}

idx = 0
for word, freq in vocab_sorted:
  if freq > 1: # 빈도수가 작은 단어는 제외
    idx += 1
    word_to_index[word] = idx

print(word_to_index)
word_to_index

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5, 'word': 6, 'keeping': 7}


{'barber': 1,
 'secret': 2,
 'huge': 3,
 'kept': 4,
 'person': 5,
 'word': 6,
 'keeping': 7}

In [90]:
# 위 결과에서 상위 5개의 단어만 사용해보자

In [91]:
vocab_size = 5

words_frequency = [  # 인덱스가 5 초과인 단어 제거
    word
    for word, idx in word_to_index.items()
    if idx > vocab_size
]

for w in words_frequency:
  del word_to_index[w]   # 해당 단어에 대한 인덱스 정보 삭제

print(word_to_index)

word_to_index

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5}


{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5}

#### OOV (Out-Of-Vocabulary)

In [92]:
# OOV 를 word_to_index  에 새롭게 추가
# 단어집합(사전) 에 없는 단어는 'OOV' 로 인덱싱 할거다.

# ['barber', 'person'] -> [1, 5]
# ['barber', 'good', 'person'] -> [1, ??, 5]   <- 단어집합에 없는 단어는 OOV 로 인코딩

In [93]:
word_to_index['OOV'] = len(word_to_index) + 1

print(word_to_index)

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5, 'OOV': 6}


In [94]:
# word_to_index  를 사용하여 sentences 의 모든 단어들을 정수 인코딩

encoded_sentences = []

for sentence in preprocessed_sentences:
  encoded_sentence = []
  for word in sentence:
    try:
      encoded_sentence.append(word_to_index[word])
    except KeyError:
      encoded_sentence.append(word_to_index['OOV'])
  encoded_sentences.append(encoded_sentence)

print(encoded_sentences)
encoded_sentences

[[1, 5], [1, 6, 5], [1, 3, 5], [6, 2], [2, 4, 3, 2], [3, 2], [1, 4, 6], [1, 4, 6], [1, 4, 2], [6, 6, 3, 2, 6, 1, 6], [1, 6, 3, 6]]


[[1, 5],
 [1, 6, 5],
 [1, 3, 5],
 [6, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [6, 6, 3, 2, 6, 1, 6],
 [1, 6, 3, 6]]

In [95]:
# 원본 문장과 인코딩된 문장을 같이 출력
for e, s in zip(encoded_sentences, sentences):
  print(f'{str(e):25} {s}')

[1, 5]                    A barber is a person.
[1, 6, 5]                 a barber is good person.
[1, 3, 5]                 a barber is huge person.
[6, 2]                    he Knew A Secret!
[2, 4, 3, 2]              The Secret He Kept is huge secret.
[3, 2]                    Huge secret.
[1, 4, 6]                 His barber kept his word.
[1, 4, 6]                 a barber kept his word.
[1, 4, 2]                 His barber kept his secret.
[6, 6, 3, 2, 6, 1, 6]     But keeping and keeping such a huge secret to himself was driving the barber crazy.
[1, 6, 3, 6]              the barber went up a huge mountain.


### Counter 사용하기

In [96]:
from collections import Counter

In [97]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [98]:
all_words_list = sum(preprocessed_sentences, [])

all_words_list

['barber',
 'person',
 'barber',
 'good',
 'person',
 'barber',
 'huge',
 'person',
 'knew',
 'secret',
 'secret',
 'kept',
 'huge',
 'secret',
 'huge',
 'secret',
 'barber',
 'kept',
 'word',
 'barber',
 'kept',
 'word',
 'barber',
 'kept',
 'secret',
 'keeping',
 'keeping',
 'huge',
 'secret',
 'driving',
 'barber',
 'crazy',
 'barber',
 'went',
 'huge',
 'mountain']

In [99]:
vocab = Counter(all_words_list)
vocab

Counter({'barber': 8,
         'person': 3,
         'good': 1,
         'huge': 5,
         'knew': 1,
         'secret': 6,
         'kept': 4,
         'word': 2,
         'keeping': 2,
         'driving': 1,
         'crazy': 1,
         'went': 1,
         'mountain': 1})

In [100]:
vocab['barber']

8

In [101]:
vocab['xxxx']

0

In [102]:
vocab_size = 5  # 사전 크기
vocab = vocab.most_common(vocab_size)  # 등장 빈도수 상위 5개 단어만

vocab

[('barber', 8), ('secret', 6), ('huge', 5), ('kept', 4), ('person', 3)]

In [103]:
# 정수 인덱스 부여

word_to_index = {}

idx = 0
for word, freq in vocab:
  idx += 1
  word_to_index[word] = idx

word_to_index


{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5}

### NLTK 의 FreqDist 사용

In [104]:
from nltk import FreqDist

In [105]:
import numpy as np
np.hstack(preprocessed_sentences)

array(['barber', 'person', 'barber', 'good', 'person', 'barber', 'huge',
       'person', 'knew', 'secret', 'secret', 'kept', 'huge', 'secret',
       'huge', 'secret', 'barber', 'kept', 'word', 'barber', 'kept',
       'word', 'barber', 'kept', 'secret', 'keeping', 'keeping', 'huge',
       'secret', 'driving', 'barber', 'crazy', 'barber', 'went', 'huge',
       'mountain'], dtype='<U8')

In [106]:
vocab = FreqDist(np.hstack(preprocessed_sentences))
vocab

FreqDist({np.str_('barber'): 8, np.str_('secret'): 6, np.str_('huge'): 5, np.str_('kept'): 4, np.str_('person'): 3, np.str_('word'): 2, np.str_('keeping'): 2, np.str_('good'): 1, np.str_('knew'): 1, np.str_('driving'): 1, ...})

In [107]:
vocab['barber']

8

In [108]:
vocab_size = 5
vocab = vocab.most_common(vocab_size)
vocab

[(np.str_('barber'), 8),
 (np.str_('secret'), 6),
 (np.str_('huge'), 5),
 (np.str_('kept'), 4),
 (np.str_('person'), 3)]

In [109]:
word_to_index = {
    word[0]: idx + 1
    for idx, word in enumerate(vocab)  # 인덱스 부여할때 enumerate 사용 하면 편리
}

word_to_index

{np.str_('barber'): 1,
 np.str_('secret'): 2,
 np.str_('huge'): 3,
 np.str_('kept'): 4,
 np.str_('person'): 5}

### Keras 의 텍스트 전처리

### Tokenizer

**tf.keras.preprocessing.text.Tokenizer**
- fit_on_texts()
- word_index
- word_counts
- texts_to_sequences()
- num_words =
- oov_token =

---

https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/text/Tokenizer

```python
tf.keras.preprocessing.text.Tokenizer(
    num_words=None,
    filters=&#x27;!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n',
    lower=True,
    split=&#x27; ',
    char_level=False,
    oov_token=None,
    analyzer=None,
    **kwargs
)
```

In [110]:
from tensorflow.keras.preprocessing.text import Tokenizer

In [111]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

#### fit_on_texts()

In [112]:
tokenizer = Tokenizer()

# fit_on_texts()안에 코퍼스를 입력으로 하면 빈도수를 기준으로 단어 집합을 생성.
tokenizer.fit_on_texts(preprocessed_sentences)


#### .word_index, .word_coutns

In [113]:
tokenizer.word_index

{'barber': 1,
 'secret': 2,
 'huge': 3,
 'kept': 4,
 'person': 5,
 'word': 6,
 'keeping': 7,
 'good': 8,
 'knew': 9,
 'driving': 10,
 'crazy': 11,
 'went': 12,
 'mountain': 13}

In [114]:
tokenizer.word_counts

OrderedDict([('barber', 8),
             ('person', 3),
             ('good', 1),
             ('huge', 5),
             ('knew', 1),
             ('secret', 6),
             ('kept', 4),
             ('word', 2),
             ('keeping', 2),
             ('driving', 1),
             ('crazy', 1),
             ('went', 1),
             ('mountain', 1)])

#### texts_to_sequences()

In [115]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [116]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[1, 5],
 [1, 8, 5],
 [1, 3, 5],
 [9, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [7, 7, 3, 2, 10, 1, 11],
 [1, 12, 3, 13]]

#### num_words=  단어사전의 크기

In [117]:
# Tokenizer (num_words=)
#   빈도수가 높은 상위 몇개의 단어만 사용 가능.


In [118]:
vocab_size = 5
tokenizer = Tokenizer(num_words = vocab_size + 1)
tokenizer.fit_on_texts(preprocessed_sentences)

In [119]:
# num_words 는 숫자를 0부터 카운트
# 만약 숫자 5를 설정하면 0 ~ 4번 단어 보존. => 1번단어 ~ 4번단어만 남는다.

# 그래서 1번 ~ 5번 단어를 사용하고 싶다면 5 + 1 값을 넣어준겁니다.

# keras Tokenizer 는 0번을 자연어처리에서 필요한 패딩(padding) 에 사용함.

In [120]:
tokenizer.word_index

{'barber': 1,
 'secret': 2,
 'huge': 3,
 'kept': 4,
 'person': 5,
 'word': 6,
 'keeping': 7,
 'good': 8,
 'knew': 9,
 'driving': 10,
 'crazy': 11,
 'went': 12,
 'mountain': 13}

In [121]:
# ↑ 일단 모든 단어에 대한 인덱스 정보는 담겨 있다
# ↓ word_counts 도 마찬가지

In [122]:
tokenizer.word_counts

OrderedDict([('barber', 8),
             ('person', 3),
             ('good', 1),
             ('huge', 5),
             ('knew', 1),
             ('secret', 6),
             ('kept', 4),
             ('word', 2),
             ('keeping', 2),
             ('driving', 1),
             ('crazy', 1),
             ('went', 1),
             ('mountain', 1)])

In [123]:
# num_words= 가 적용되는 것은 texts_to_sequences() 를 사용할때 적용됨!

In [124]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[1, 5],
 [1, 5],
 [1, 3, 5],
 [2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4],
 [1, 4],
 [1, 4, 2],
 [3, 2, 1],
 [1, 3]]

In [125]:
# ↑상위 5개 단어만 (1번~5번) 보존되고 나머지 단어들은 제거된 것과 같은 결과.

"""
[[1, 5],      <-   ['barber', 'person'],
 [1, 5],      <-   ['barber', 'good', 'person']  num_words= 바깥의 단어 'good' 는 제거된 결과
 [1, 3, 5],
 [2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4],
 [1, 4],
 [1, 4, 2],
 [3, 2, 1],
 [1, 3]]
코퍼스에 대해서 각 단어를 이미 정해진 인덱스로 변환하는데,
상위 5개의 단어만을 사용하겠다고 지정하였으므로 1번 단어부터 5번 단어까지만 보존되고
num_words= 바깥의 단어 단어들은 제거된 것을 볼 수 있습니다.
"""


"\n[[1, 5],      <-   ['barber', 'person'],\n [1, 5],      <-   ['barber', 'good', 'person']  num_words= 바깥의 단어 'good' 는 제거된 결과\n [1, 3, 5],\n [2],\n [2, 4, 3, 2],\n [3, 2],\n [1, 4],\n [1, 4],\n [1, 4, 2],\n [3, 2, 1],\n [1, 3]]\n코퍼스에 대해서 각 단어를 이미 정해진 인덱스로 변환하는데,\n상위 5개의 단어만을 사용하겠다고 지정하였으므로 1번 단어부터 5번 단어까지만 보존되고\nnum_words= 바깥의 단어 단어들은 제거된 것을 볼 수 있습니다.\n"

#### oov_token=
- 단어집합에 없는 단어들 (out of vocabulary)
- keras 는 'OOV' 인덱스를 1 로 설정 (기본)

In [126]:
# 인덱스 0 (PAD), 1 (OOV) 고려하여 단어집합 크기는 +2
# 실제 단어는 인덱스 2부터 시작
vocab_size = 5
tokenizer = Tokenizer(num_words = vocab_size + 2, oov_token='OOV')
tokenizer.fit_on_texts(preprocessed_sentences)

In [127]:
tokenizer.word_index

{'OOV': 1,
 'barber': 2,
 'secret': 3,
 'huge': 4,
 'kept': 5,
 'person': 6,
 'word': 7,
 'keeping': 8,
 'good': 9,
 'knew': 10,
 'driving': 11,
 'crazy': 12,
 'went': 13,
 'mountain': 14}

In [128]:
tokenizer.index_word

{1: 'OOV',
 2: 'barber',
 3: 'secret',
 4: 'huge',
 5: 'kept',
 6: 'person',
 7: 'word',
 8: 'keeping',
 9: 'good',
 10: 'knew',
 11: 'driving',
 12: 'crazy',
 13: 'went',
 14: 'mountain'}

In [129]:
tokenizer.word_index['OOV']

1

In [130]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[2, 6],
 [2, 1, 6],
 [2, 4, 6],
 [1, 3],
 [3, 5, 4, 3],
 [4, 3],
 [2, 5, 1],
 [2, 5, 1],
 [2, 5, 3],
 [1, 1, 4, 3, 1, 2, 1],
 [2, 1, 4, 1]]

In [131]:
"""
[[2, 6],      <-   ['barber', 'person'],
 [2, 1, 6],    <-   ['barber', 'good', 'person']
 [2, 4, 6],
 [1, 3],
 [3, 5, 4, 3],
 [4, 3],
 [2, 5, 1],
 [2, 5, 1],
 [2, 5, 3],
 [1, 1, 4, 3, 1, 2, 1],
 [2, 1, 4, 1]]

 빈도수 상위 5개 단어 2 ~ 6 까지의 인덱스.
 단어집합에 없는 단어는 'OOV' 인 인덱스 1 로 인코딩
"""
None

### HuggingFace Tokenizer / Transformers

In [132]:
# BERT, GPT, RoBERTa 같은 '사전학습된 모델'을 쓴다면,
# Hugging Face의 transformers 라이브러리가 더 강력하고 표준적인 선택입니다.

# 장점!
#  - WordPiece/BPE 토크나이징 자동 지원
#  - 패딩, 마스크, 어텐션 생성까지 자동
#  - PyTorch 텐서 바로 반환
# 👉 Keras보다 훨씬 강력하고 범용적인 방법입니다.

In [133]:
from transformers import AutoTokenizer

In [134]:
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')

In [135]:
text = ['I love Python', "It is great"]

encodings = tokenizer(text, padding=True, truncation=True, return_tensors='pt')

from pprint import pprint
pprint(encodings)


{'attention_mask': tensor([[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]]),
 'input_ids': tensor([[  101,  1045,  2293, 18750,   102],
        [  101,  2009,  2003,  2307,   102]]),
 'token_type_ids': tensor([[0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]])}


In [136]:
"""
{'attention_mask': tensor([[1, 1, 1, 1, 1],
                          [1, 1, 1, 1, 1]]),
 'input_ids': tensor([[  101,  1045,  2293, 18750,   102],
                      [  101,  2009,  2003,  2307,   102]]),
 'token_type_ids': tensor([[0, 0, 0, 0, 0],
                          [0, 0, 0, 0, 0]])}
"""

encodings.input_ids  # 토큰의 정수 인덱싱

tensor([[  101,  1045,  2293, 18750,   102],
        [  101,  2009,  2003,  2307,   102]])

In [137]:
encoding = tokenizer("I love python", "It is great")
pprint(encoding['token_type_ids'])

[0, 0, 0, 0, 0, 1, 1, 1, 1]


In [138]:
pprint(encoding)

{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1],
 'input_ids': [101, 1045, 2293, 18750, 102, 2009, 2003, 2307, 102],
 'token_type_ids': [0, 0, 0, 0, 0, 1, 1, 1, 1]}


In [139]:
encoding = tokenizer(['I love python', 'love'], padding=True)

pprint(encoding)

{'attention_mask': [[1, 1, 1, 1, 1], [1, 1, 1, 0, 0]],
 'input_ids': [[101, 1045, 2293, 18750, 102], [101, 2293, 102, 0, 0]],
 'token_type_ids': [[0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]}


### special tokens
`AutoTokenizer` (예: BERT 계열 등)에는 **모델 내부적으로 특별한 의미를 가지는 “스페셜 토큰(special tokens)”** 이 여러 개 정의되어 있다.
이 토큰들은 단순한 단어가 아니라, **모델이 문장의 구조를 이해하기 위해 꼭 필요한 신호 역할**을 한다.

---

## 🧩 대표적인 Special Tokens (BERT 기준)

| 토큰                          | 예시 ID (bert-base-uncased 기준) | 의미              | 역할                                |
| --------------------------- | ---------------------------- | --------------- | --------------------------------- |
| `[PAD]`                     | `0`                          | Padding Token   | 짧은 문장 길이 맞추기용                     |
| `[UNK]`                     | `100`                        | Unknown Token   | 어휘(vocab)에 없는 단어                  |
| `[CLS]`                     | `101`                        | Class Token     | 문장 전체 표현 (classification용)        |
| `[SEP]`                     | `102`                        | Separator Token | 문장 구분 (문장 A/B 구분)                 |
| `[MASK]`                    | `103`                        | Mask Token      | Masked Language Modeling에서 가려진 단어 |
| `[unused0]` ~ `[unused999]` | (104~)                       | 미사용 토큰          | 예비 슬롯 (특수 태그용으로 종종 활용)            |

> 💡 **ID 값은 모델마다 다를 수 있습니다!**
> 예: RoBERTa나 DistilBERT는 `[CLS]`가 `<s>`로, `[SEP]`가 `</s>`로 이름이 다릅니다.

---

## 🔍 각 토큰의 구체적 역할

### 1️⃣ `[CLS]` (Classify)

* 문장 맨 앞에 항상 추가됨
* 문장 전체의 의미를 요약하는 “대표 토큰”
* 분류(Classification) 작업 시 이 토큰의 벡터를 사용

**예시:**

```python
[CLS] I love PyTorch [SEP]
```

---

### 2️⃣ `[SEP]` (Separator)

* 문장 끝 혹은 문장 사이에 추가됨
* 문장 경계를 표시하거나 문장쌍(A/B) 구분용으로 사용

**예시:**

```python
[CLS] I love PyTorch [SEP] It is great [SEP]
```

---

### 3️⃣ `[PAD]`

* 배치 입력 시 문장 길이를 맞추기 위해 뒤쪽에 채워짐
* 모델이 무시하도록 `attention_mask`에서 0으로 표시됨

```python
[CLS] I love PyTorch [SEP] [PAD] [PAD]
```

---

### 4️⃣ `[UNK]`

* 토크나이저의 어휘 사전에 없는 단어가 들어왔을 때 사용

```python
"I love 🦖" → [CLS] I love [UNK] [SEP]
```

---

### 5️⃣ `[MASK]`

* 사전학습(Pretraining) 중 일부 단어를 가려서 모델이 예측하게 만듦
* Masked Language Modeling(MLM)에서 사용

```python
[CLS] I love [MASK] [SEP]
```

모델은 `[MASK]` 위치의 단어를 예측합니다 → `"PyTorch"`

---


## 📚 모델별 Special Token 비교

| 모델                          | 문장 시작   | 문장 종료   | 패딩      | 마스크              | 비고                 |
| --------------------------- | ------- | ------- | ------- | ---------------- | ------------------ |
| **BERT**                    | `[CLS]` | `[SEP]` | `[PAD]` | `[MASK]`         | 전통적인 구조            |
| **RoBERTa**                 | `<s>`   | `</s>`  | `<pad>` | `<mask>`         | BERT 변형            |
| **DistilBERT**              | `[CLS]` | `[SEP]` | `[PAD]` | `[MASK]`         | BERT 경량화           |
| **GPT-2 / GPT-3 / GPT-Neo** | 없음      | ``      | 없음      | 없음               | 단방향 언어모델           |
| **T5**                      | `<pad>` | `</s>`  | `<pad>` | `<extra_id_0>` 등 | Encoder-Decoder 구조 |

---

## 🧠 요약

| 토큰       | 용도           | 일반적 ID (BERT 기준) |
| -------- | ------------ | ---------------- |
| `[PAD]`  | 패딩           | `0`              |
| `[CLS]`  | 문장 대표        | `101`            |
| `[SEP]`  | 문장 구분        | `102`            |
| `[MASK]` | 단어 가리기 (MLM) | `103`            |
| `[UNK]`  | 미등록 단어       | `100`            |



In [140]:
tokenizer.special_tokens_map

{'unk_token': '[UNK]',
 'sep_token': '[SEP]',
 'pad_token': '[PAD]',
 'cls_token': '[CLS]',
 'mask_token': '[MASK]'}

# ■ 02-07 패딩(Padding)

In [141]:
# 자연어 처리를 하다보면 각 문장(또는 문서)은 서로 길이가 다를 수 있습니다.
# 그런데 기계는 길이가 전부 동일한 문서들에 대해서는 하나의 행렬로 보고,
# 한꺼번에 묶어서 처리할 수 있습니다.

# 다시 말해 병렬 연산을 위해서 여러 문장의 길이를 임의로 동일하게 맞춰주는 작업이 필요!.

![](https://miro.medium.com/v2/resize:fit:1400/1*CPLhZoVSTCWgAxe2LKXoOA.png)

## Numpy 로 패딩하기

In [142]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [143]:
# 단어 집합을 만들고, 정수 인코딩을 수행합니다.

tokenizer = Tokenizer()
# fit_on_texts()안에 코퍼스를 입력으로 하면 빈도수를 기준으로 단어 집합을 생성.
tokenizer.fit_on_texts(preprocessed_sentences)

encoded = tokenizer.texts_to_sequences(preprocessed_sentences)
print(encoded)

# ↓ 모든 단어가 고유한 정수로 변환되었습니다.

[[1, 5], [1, 8, 5], [1, 3, 5], [9, 2], [2, 4, 3, 2], [3, 2], [1, 4, 6], [1, 4, 6], [1, 4, 2], [7, 7, 3, 2, 10, 1, 11], [1, 12, 3, 13]]


In [144]:
# 모든 문장을 동일한 길이로 맞춰주기 위해서 이 중에서 가장 길이가 긴 문장의 길이를 계산
max_len = max(len(item) for item in encoded)
max_len

7

In [145]:
# 모든 문장의 길이를 max_len 으로 맞추어봅니다
# 패딩문자는 'PAD' 라 하고, 정수 0 으로 정의
# 길이가 7보다 짧은 문장에는 숫자 0 으로 채워짐.

In [146]:
for sentence in encoded:  # 각 문장에 대해서
  while len(sentence) < max_len:
    sentence.append(0)   # 0 'PAD' 을 뒤에 추가

np.array(encoded)


array([[ 1,  5,  0,  0,  0,  0,  0],
       [ 1,  8,  5,  0,  0,  0,  0],
       [ 1,  3,  5,  0,  0,  0,  0],
       [ 9,  2,  0,  0,  0,  0,  0],
       [ 2,  4,  3,  2,  0,  0,  0],
       [ 3,  2,  0,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  2,  0,  0,  0,  0],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 1, 12,  3, 13,  0,  0,  0]])

## keras 의 pad_seqeunces()

In [147]:
from keras.utils import pad_sequences

https://www.tensorflow.org/api_docs/python/tf/keras/utils/pad_sequences
```python
tf.keras.utils.pad_sequences(
    sequences,
    maxlen=None,
    dtype='int32',
    padding='pre',
    truncating='pre',
    value=0.0
)
```

In [148]:
encoded = tokenizer.texts_to_sequences(preprocessed_sentences)
print(encoded)

[[1, 5], [1, 8, 5], [1, 3, 5], [9, 2], [2, 4, 3, 2], [3, 2], [1, 4, 6], [1, 4, 6], [1, 4, 2], [7, 7, 3, 2, 10, 1, 11], [1, 12, 3, 13]]


In [149]:
padded = pad_sequences(encoded)
padded

array([[ 0,  0,  0,  0,  0,  1,  5],
       [ 0,  0,  0,  0,  1,  8,  5],
       [ 0,  0,  0,  0,  1,  3,  5],
       [ 0,  0,  0,  0,  0,  9,  2],
       [ 0,  0,  0,  2,  4,  3,  2],
       [ 0,  0,  0,  0,  0,  3,  2],
       [ 0,  0,  0,  0,  1,  4,  6],
       [ 0,  0,  0,  0,  1,  4,  6],
       [ 0,  0,  0,  0,  1,  4,  2],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 0,  0,  0,  1, 12,  3, 13]], dtype=int32)

In [150]:
padded = pad_sequences(encoded, padding='post')
padded

array([[ 1,  5,  0,  0,  0,  0,  0],
       [ 1,  8,  5,  0,  0,  0,  0],
       [ 1,  3,  5,  0,  0,  0,  0],
       [ 9,  2,  0,  0,  0,  0,  0],
       [ 2,  4,  3,  2,  0,  0,  0],
       [ 3,  2,  0,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  2,  0,  0,  0,  0],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 1, 12,  3, 13,  0,  0,  0]], dtype=int32)

In [151]:
padded = pad_sequences(encoded, padding='post', maxlen=5)
padded

array([[ 1,  5,  0,  0,  0],
       [ 1,  8,  5,  0,  0],
       [ 1,  3,  5,  0,  0],
       [ 9,  2,  0,  0,  0],
       [ 2,  4,  3,  2,  0],
       [ 3,  2,  0,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  2,  0,  0],
       [ 3,  2, 10,  1, 11],
       [ 1, 12,  3, 13,  0]], dtype=int32)

In [152]:
padded = pad_sequences(encoded, padding='post', truncating='post', maxlen=5)
padded

array([[ 1,  5,  0,  0,  0],
       [ 1,  8,  5,  0,  0],
       [ 1,  3,  5,  0,  0],
       [ 9,  2,  0,  0,  0],
       [ 2,  4,  3,  2,  0],
       [ 3,  2,  0,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  2,  0,  0],
       [ 7,  7,  3,  2, 10],
       [ 1, 12,  3, 13,  0]], dtype=int32)

In [153]:
tokenizer.word_index

{'barber': 1,
 'secret': 2,
 'huge': 3,
 'kept': 4,
 'person': 5,
 'word': 6,
 'keeping': 7,
 'good': 8,
 'knew': 9,
 'driving': 10,
 'crazy': 11,
 'went': 12,
 'mountain': 13}

In [154]:
padded = pad_sequences(encoded, padding='post', value=len(tokenizer.word_index) + 1)
padded

array([[ 1,  5, 14, 14, 14, 14, 14],
       [ 1,  8,  5, 14, 14, 14, 14],
       [ 1,  3,  5, 14, 14, 14, 14],
       [ 9,  2, 14, 14, 14, 14, 14],
       [ 2,  4,  3,  2, 14, 14, 14],
       [ 3,  2, 14, 14, 14, 14, 14],
       [ 1,  4,  6, 14, 14, 14, 14],
       [ 1,  4,  6, 14, 14, 14, 14],
       [ 1,  4,  2, 14, 14, 14, 14],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 1, 12,  3, 13, 14, 14, 14]], dtype=int32)

## HuggingFace Tokenizer / Transformers

In [155]:
# transformers.AutoTokenizer의 padding 과 truncation 옵션은
# 토크나이징 시 입력 길이를 맞추기 위해 사용됨.
# 이 두 가지는 특히 배치(batch) 입력 시 매우 중요

In [156]:
# padding=
# 문장 길이 맞추기 (짧은 문장 뒤에 0을 채움)

# 모든 문장이 길이가 다르기 때문에, 배치 텐서로 묶기 전에 길이를 통일해야 합니다.
# 짧은 문장에는 [PAD] 토큰(보통 id=0)을 추가해서 길이를 맞춥니다.

In [157]:
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')

In [158]:
batch = tokenizer(
    ["I love PyTorch", "PyTorch is great and lovely"],
    padding=True,
    return_tensors="pt"
)

batch['input_ids']

tensor([[  101,  1045,  2293,  1052, 22123,  2953,  2818,   102,     0,     0],
        [  101,  1052, 22123,  2953,  2818,  2003,  2307,  1998,  8403,   102]])

In [159]:
batch = tokenizer(
    ["I love PyTorch", "PyTorch is great and lovely"],
    padding='max_length',
    max_length=10,
    return_tensors="pt"
)

batch['input_ids']

tensor([[  101,  1045,  2293,  1052, 22123,  2953,  2818,   102,     0,     0],
        [  101,  1052, 22123,  2953,  2818,  2003,  2307,  1998,  8403,   102]])

In [160]:
batch = tokenizer(
    "I Love PyTorch" * 100,
    truncation=True,
    max_length=10,
    return_tensors="pt"
)

batch['input_ids']

tensor([[  101,  1045,  2293,  1052, 22123,  2953,  5428,  2293,  1052,   102]])

# ■ 원-핫 인코딩 (One-Hot Encoding)

## 단어 집합 (Vocabulary)

In [161]:
# 원-핫 인코딩을 위해서 먼저 해야할 일은 '단어 집합'을 만드는 일입니다.
# 텍스트의 모든 단어를 '중복을 허용하지 않고' 모아놓으면 이를 단어 집합

# 그리고 이 단어 집합에 고유한 정수를 부여하는 정수 인코딩을 진행합니다.
# 텍스트에 단어가 총 5,000개가 존재한다면, 단어 집합의 크기는 5,000입니다.

# 각 단어에 고유한 정수 인덱싱이 되었다고 할때
# 이 '숫자'로 바뀐 단어들을 '벡터' 로 다루고 싶다면?


![](https://heung-bae-lee.github.io/image/why_is_sparse_matirx_one_hot_encoding.png)

## One-hot encoding 작성 (수동)

In [162]:
# 단어집합의 크기를 벡터의 '차원' 으로 하고
# 표현하고 싶은 단어의 인덱스에 '1' 값을 부여, 다른 인덱스에는 '0' 을 부여하여 단어의 벡터 표현
# 이렇게 표현된 벡터를 원-핫 벡터 (one-hot vector)


In [163]:
tokens = okt.morphs("나는 자연어 처리를 배운다")

print(tokens)

['나', '는', '자연어', '처리', '를', '배운다']


In [164]:
# 각 토큰에 대해서 고유한 정수를 부여합니다.
# 지금은 문장이 짧기 때문에 각 단어의 빈도수를 고려하지 않지만,
# 빈도수 순으로 단어를 정렬하여 정수를 부여하는 경우가 많습니다.

word_to_index = {word : index for index, word in enumerate(tokens)}
print('단어 집합 :',word_to_index)



단어 집합 : {'나': 0, '는': 1, '자연어': 2, '처리': 3, '를': 4, '배운다': 5}


In [165]:
def ont_hot_encoding(word, word_to_index):
  one_hot_vector = [0] * (len(word_to_index))  # 단어집합의 크기를 벡터의 '차원크기' 로 하고
  index = word_to_index[word]   #
  one_hot_vector[index] = 1  # 표현하고 싶은 단어의 인덱스에 '1' 값을 부여
  return one_hot_vector

In [166]:
ont_hot_encoding("자연어", word_to_index)

[0, 0, 1, 0, 0, 0]

## keras 의 to_categorical()

In [167]:
from keras.utils import to_categorical

https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical
```python
tf.keras.utils.to_categorical(
    x, num_classes=None
)
```

In [168]:
text = "나랑 점심 먹으러 갈래 점심 메뉴는 햄버거 갈래 갈래 햄버거 최고야"
# 위와 같은 문장이 있다고 했을 때, 케라스 토크나이저를 이용한 정수 인코딩은 다음과 같습니다.
tokenizer = Tokenizer()

tokenizer.fit_on_texts([text])

print('단어집합: ', tokenizer.word_index)



단어집합:  {'갈래': 1, '점심': 2, '햄버거': 3, '나랑': 4, '먹으러': 5, '메뉴는': 6, '최고야': 7}


In [169]:
sub_text = "점심 먹으러 갈래 메뉴는 햄버거 최고야"

In [171]:
encoded = tokenizer.texts_to_sequences([sub_text])[0]
encoded

[2, 5, 1, 6, 3, 7]

In [173]:
# 위 정수 인코딩 된 결과를 one-hot encoding 하기
one_hot = to_categorical(encoded)
print(one_hot)

# 사전(단어집합)의 크기는 7
# 토큰의 개수는 6개
# => 6 x 7 벡터


[[0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1.]]


In [ ]:
"""
[[0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1.]]
"""
None

In [174]:
tokenizer.index_word

{1: '갈래', 2: '점심', 3: '햄버거', 4: '나랑', 5: '먹으러', 6: '메뉴는', 7: '최고야'}

## PyTorch 의 one_hot()

In [177]:
import torch.nn.functional as F

In [178]:
y = torch.tensor([0, 2, 1])

y_onehot = F.one_hot(y, num_classes=3)  # 기본적을 int형 tensor 다

y_onehot

tensor([[1, 0, 0],
        [0, 0, 1],
        [0, 1, 0]])

In [179]:
F.one_hot(y).float()  # 모델 입력으로 사용하려면 float() 변환!

tensor([[1., 0., 0.],
        [0., 0., 1.],
        [0., 1., 0.]])

##  원-핫 인코딩(One-Hot Encoding)의 한계
- 저장공간의 문제
- 단어의 유사도 표현 못함

In [ ]:
# 저장공간 문제]
# 원핫벡터에서 '단어집합의 크기' 가 곧 벡터의 차원크기가 된다!
# 즉 단어개수가 늘어날수록 벡터를 저장하기 위한 공긴이 급속이 늘어난다

In [ ]:
# 단어의 유사성 알수 있다면..
#  웹 검색창에 '오사카 숙소' 검색.
#   '숙소' 와 유사한 결과들을 낼수 있다  ex) 호텔, 게스트 하우스, 료칸...

# 원핫벡터는 단어의 유사도를 표현하지 못함...

In [ ]:
# 예를 들어서 늑대, 호랑이, 강아지, 고양이라는 4개의 단어에 대해서 원-핫 인코딩을 해서
# 각각, [1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]이라는 원-핫 벡터를 부여받았다고 합시다.
# 이때 원-핫 벡터로는 강아지와 늑대가 유사하고, 호랑이와 고양이가 유사하다는 것을 표현할 수가 없습니다.


In [ ]:
# 단어의 유사도 문제 해결 -> 단어를 다차원 공간에 벡터화 기법 두가지
# 1. 카운트 기반의 벡터화 : ex) LSA(잠재 의미 분석), HAL
# 2. 예측 기반으로 벡터화 : NNLM, RNNLM, Word2Vec, FastText
#  카운트 기반과 예측 기반 두 가지 방법을 모두 사용 : ex) GloVe


# ■ Embedding

![](https://heung-bae-lee.github.io/image/why_is_sparse_matirx_one_hot_encoding.png)

![](https://heung-bae-lee.github.io/image/Dense_representation.png)
**Embedding Layer**

![](https://blog.kakaocdn.net/dna/LA1OL/btsMcg2UGGC/AAAAAAAAAAAAAAAAAAAAAF-I8ptCXJ_ZEsKoyFiBI7tpX-mxzv7YuQmachNEkty_/img.jpg?credential=yqXZFxpELC7KVnFOS48ylbz2pIh7yKj8&expires=1761922799&allow_ip=&allow_referer=&signature=h9CoTgqrvIpkaFM6RY9%2FVB3zitk%3D)

In [ ]:
# 자연어 처리(Natural Language Processing)분야에서 임베딩(Embedding)은
# 사람이 쓰는 자연어를 기계가 이해할 수 있는 숫자형태인 vector로 바꾼 결과 혹은 그 일련의 과정 전체를 의미한다.

## PyTorch의 nn.Embedding()

**torch.nn.Embedding**
https://docs.pytorch.org/docs/stable/generated/torch.nn.Embedding.html

```python
class torch.nn.Embedding(
  num_embeddings,
  embedding_dim,
  padding_idx=None,
  max_norm=None,
  norm_type=2.0,
  scale_grad_by_freq=False,
  sparse=False,
  _weight=None,
  _freeze=False,
  device=None,
  dtype=None)
```
None


In [180]:
# 단어 사전 크기 (vocab_size) 와 임베딩 차원 설정
vocab_size = 10   # 단어집합에 10개 있다고 가정 (0 ~ 9)
embedding_dim = 4  # 각 단어를 4차원 벡터로 표현.

In [182]:
import torch
import torch.nn as nn

# Embedding 레이어 생성
embedding = nn.Embedding(num_embeddings=vocab_size, embedding_dim=embedding_dim)

In [183]:
# 입력으론 단어 인덱스(정수)를 전달
# 문장 [2, 5, 7]
input_indices = torch.tensor([2, 5, 7])

input_indices

tensor([2, 5, 7])

In [184]:
# 임베딩 결과 출력
embedded = embedding(input_indices)

print("입력 인덱스:", input_indices)
print("임베딩 결과 텐서 크기:", embedded.shape)
  # 결과 텐서 크기 [입력길이, embedding_dim]  즉, [3, 4] 형태
print("임베딩 벡터:\n", embedded)

입력 인덱스: tensor([2, 5, 7])
임베딩 결과 텐서 크기: torch.Size([3, 4])
임베딩 벡터:
 tensor([[-0.2652,  0.3334,  0.2629, -0.2879],
        [ 1.4292, -0.5325,  0.6355, -0.3047],
        [-0.4224,  0.2216, -0.6401,  1.4858]], grad_fn=<EmbeddingBackward0>)


In [189]:
embedded.detach().numpy()

array([[-0.26518002,  0.3333805 ,  0.26289576, -0.28790477],
       [ 1.4291832 , -0.53253275,  0.6355242 , -0.30469966],
       [-0.42242065,  0.2216359 , -0.6400572 ,  1.4857645 ]],
      dtype=float32)

In [191]:
# batch 로 입력

input_batch = torch.tensor([[1, 2, 3],
                            [4, 5, 6]])

embedded_batch = embedding(input_batch)

print(embedded_batch.shape)

# 배치2개 (문장2개), 각 문장(시퀀스)의 단어 3단어. 임베딩 차원 4
# [batch_size, sentence_length, embedding_dim]



torch.Size([2, 3, 4])


In [197]:
# 단어사전이 주어진 경우.
vocab = {"I": 0, "love": 1, "AI": 2, "hate": 3, "you": 4}
vocab_size = len(vocab)
embedding_dim = 5  # 각 단어를 5차원 벡터로 표현

embedding = nn.Embedding(vocab_size, embedding_dim)

# 문장을 단어 인덱스로 변환
sentence = ["I", "love", "AI"]

input_indices = torch.tensor([vocab[word] for word in sentence])
print('입력 문장:', sentence)
print('단어 인덱스:', input_indices)

# 임베딩 결과
embedded = embedding(input_indices)
print("\n임베딩 결과 텐서 크기:", embedded.shape)  # 3 x 5
print("임베딩 벡터:\n", embedded)

입력 문장: ['I', 'love', 'AI']
단어 인덱스: tensor([0, 1, 2])

임베딩 결과 텐서 크기: torch.Size([3, 5])
임베딩 벡터:
 tensor([[ 0.8016,  2.3784,  0.1821,  1.5165, -0.6410],
        [-0.9772, -0.0480, -0.9042, -0.8174, -1.4398],
        [-1.7991,  0.5025, -0.0130,  0.2589, -0.0176]],
       grad_fn=<EmbeddingBackward0>)


In [198]:
# 추가: 배치 입력 예시
# 2개의 문장 배치
sentences = [
    ["I", "love", "AI"],
    ["I", "hate", "you"]
]

# 단어 인덱스로 변환
input_batch = torch.tensor([[vocab[word] for word in sent] for sent in sentences])

# (batch_size=2, seq_len=3)
embedded_batch = embedding(input_batch)

print("\n배치 임베딩 크기:", embedded_batch.shape)



배치 임베딩 크기: torch.Size([2, 3, 5])


## keras 의 Embedding

**tf.keras.layers.Embedding**

https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding

```python
tf.keras.layers.Embedding(
    input_dim,
    output_dim,
    embeddings_initializer='uniform',
    embeddings_regularizer=None,
    embeddings_constraint=None,
    mask_zero=False,
    weights=None,
    lora_rank=None,
    **kwargs
)
```

In [199]:
import tensorflow as tf

# 1️⃣ 단어 사전 크기(vocab_size)와 임베딩 차원 설정
vocab_size = 10     # 단어가 10개 (0~9)
embedding_dim = 4   # 각 단어를 4차원 벡터로 표현

# 2️⃣ Embedding 레이어 생성
embedding = tf.keras.layers.Embedding(input_dim=vocab_size, output_dim=embedding_dim)

# 3️⃣ 입력 (정수 인덱스)
# 문장: [2, 5, 7]
input_indices = tf.constant([2, 5, 7])

# 4️⃣ 임베딩 벡터 조회
embedded = embedding(input_indices)

print("입력 인덱스:", input_indices.numpy())
print("임베딩 결과 텐서 shape:", embedded.shape)
print("임베딩 벡터:\n", embedded.numpy())
# 출력 예시 (매번 달라짐)

입력 인덱스: [2 5 7]
임베딩 결과 텐서 shape: (3, 4)
임베딩 벡터:
 [[ 0.03394177  0.02249842  0.04363859 -0.04641441]
 [-0.02981851 -0.00140935  0.04017014  0.01292807]
 [-0.01267532  0.01427866 -0.04492049 -0.04461707]]


In [200]:
# 배치 입력

# 두 개의 문장 (batch size = 2, 각 문장 길이 = 3)
input_batch = tf.constant([
    [1, 2, 3],
    [4, 5, 6]
])

embedded_batch = embedding(input_batch)

print("입력 shape:", input_batch.shape)
print("출력 shape:", embedded_batch.shape)

입력 shape: (2, 3)
출력 shape: (2, 3, 4)


In [208]:
# 1️⃣ 단어 사전 정의
vocab = {"I": 0, "love": 1, "AI": 2, "hate": 3, "you": 4}
vocab_size = len(vocab)
embedding_dim = 5  # 각 단어를 5차원 벡터로 표현

# 2️⃣ Embedding 레이어 생성
embedding = tf.keras.layers.Embedding(input_dim=vocab_size, output_dim=embedding_dim)

# 3️⃣ 문장을 단어 인덱스로 변환
sentence = ["I", "love", "AI"]
input_indices = tf.constant([vocab[word] for word in sentence])

print("입력 문장:", sentence)
print("단어 인덱스:", input_indices.numpy())

# 4️⃣ 임베딩 결과 얻기
embedded = embedding(input_indices)

print("\n임베딩 결과 텐서 shape:", embedded.shape)
print("임베딩 벡터:\n", embedded.numpy())

입력 문장: ['I', 'love', 'AI']
단어 인덱스: [0 1 2]

임베딩 결과 텐서 shape: (3, 5)
임베딩 벡터:
 [[-0.03217633  0.03995171 -0.01169977 -0.02028978 -0.00827318]
 [ 0.04088366  0.02004148  0.00165981 -0.02272731 -0.04116873]
 [ 0.03508634 -0.02344067  0.02167862 -0.04724714 -0.03886967]]


In [209]:
# 2개의 문장 배치
sentences = [
    ["I", "love", "AI"],
    ["I", "hate", "you"]
]

# 단어를 인덱스로 변환
input_batch = tf.constant([[vocab[w] for w in s] for s in sentences])

# (batch=2, seq_len=3)
embedded_batch = embedding(input_batch)

print("\n입력 shape:", input_batch.shape)
print("출력 shape:", embedded_batch.shape)


입력 shape: (2, 3)
출력 shape: (2, 3, 5)
